### PR TITLE
Remove all globbed dist-infos in editable self-installs

### PR DESCRIPTION
**Pull request by nickray**
# Pull Request Check List

Resolves: https://github.com/python-poetry/poetry/issues/2918, https://github.com/python-poetry/poetry/issues/3741, partially https://github.com/python-poetry/poetry/issues/2855.

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

Depends on: https://github.com/python-poetry/poetry-core/pull/137

The `SitePackages` class seems not quite made for globbing. Appreciate pointers if there is a better way..

This depends on https://github.com/python-poetry/poetry-core/pull/137 to avoid code duplication, might be easier to just implement the method from there here.

### CODE REVIEW

**Comment by nickray** [comment]
This was mostly good to go, until a change in poetry-core/master broke the poetry/master build. How to proceed?

File: poetry/masonry/builders/editable.py [hunk 1]
```diff
@@ -197,21 +197,27 @@ def _add_dist_info(self, added_files: List[Path]) -> None:
...
-        ):
-            if dist_info.exists():
-                self._debug(
-                    "  - Removing existing <c2>{}</c2> directory from <b>{}</b>".format(
-                        dist_info.name, dist_info.parent
+        # remove dist_info for all existing versions, in all candidate directories we can modify
+        dist_info_paths = [
+            Path(dist_info)
+            for site_package_dir in self._env.site_packages.writable_candidates
+            for dist_info in site_package_dir.glob(builder.dist_info_glob)
```
Comment (abn) [inline poetry/masonry/builders/editable.py:206]: Since the glob is only really required for an editable builder; you can construct it here.

**Review by abn — CHANGES_REQUESTED** [review]
You could add a `find_glob` method to `SitePackages`. That might be the clean way forward.

File: poetry/masonry/builders/editable.py (hunk 1 above)
Comment (nickray) [reply poetry/masonry/builders/editable.py:206]: I can, sure. If I do, though, `return "{}-*.dist-info".format(escaped_name)` depends on implementation details (structure of format string, implementation of `escape_name(self._package.name)`), which I felt was iffy.

**Comment by fredrikaverpil** [comment]
This issue is of interest to me, is there anything I can do to help?

**Comment by fredrikaverpil** [comment]
@stephsamson @abn maybe this should be closed now that #3900 was merged?

**Comment by abn** [comment]
Superseded by #3900.

Thanks @fredrikaverpil

**Pull request closed without merging**